core/unix/coredump.c

```diff
@@ -154,6 +194,178 @@ write_section_header(DR_PARAM_IN file_t elf_file,
+    regs->r14 = mcontext->r14;
+    regs->r15 = mcontext->r15;
+    regs->rip = (uint64_t)mcontext->rip;
+    regs->eflags = mcontext->rflags;
+#elif defined(AARCH64)
+    regs->regs[0] = mcontext->r0;
+    regs->regs[1] = mcontext->r1;
+    regs->regs[2] = mcontext->r2;
+    regs->regs[3] = mcontext->r3;
+    regs->regs[4] = mcontext->r4;
+    regs->regs[5] = mcontext->r5;
+    regs->regs[6] = mcontext->r6;
+    regs->regs[7] = mcontext->r7;
+    regs->regs[8] = mcontext->r8;
+    regs->regs[9] = mcontext->r9;
+    regs->regs[10] = mcontext->r10;
+    regs->regs[11] = mcontext->r11;
+    regs->regs[12] = mcontext->r12;
+    regs->regs[13] = mcontext->r13;
+    regs->regs[14] = mcontext->r14;
+    regs->regs[15] = mcontext->r15;
+    regs->regs[16] = mcontext->r16;
+    regs->regs[17] = mcontext->r17;
+    regs->regs[18] = mcontext->r18;
+    regs->regs[19] = mcontext->r19;
+    regs->regs[20] = mcontext->r20;
+    regs->regs[21] = mcontext->r21;
+    regs->regs[22] = mcontext->r22;
+    regs->regs[23] = mcontext->r23;
+    regs->regs[24] = mcontext->r24;
+    regs->regs[25] = mcontext->r25;
+    regs->regs[26] = mcontext->r26;
+    regs->regs[27] = mcontext->r27;
+    regs->regs[28] = mcontext->r28;
+    regs->regs[29] = mcontext->r29;
+    regs->regs[30] = mcontext->r30;
+    regs->sp = mcontext->sp;
+    regs->pc = (uint64_t)mcontext->pc;
+#endif
+}
+
+/*
+ * Write prstatus structure to the file. Returns true if the note is written to
+ * the file, false otherwise.
+ */
+static bool
+write_prstatus_note(DR_PARAM_IN file_t elf_file)
+{
+    void *drcontext = dr_get_current_drcontext();
+    dr_mcontext_t mcontext;
+    mcontext.size = sizeof(mcontext);
+    mcontext.flags = DR_MC_ALL;
+    dr_get_mcontext(drcontext, &mcontext);
+
+    struct elf_prstatus prstatus;
+    struct user_regs_struct *regs = (struct user_regs_struct *)&prstatus.pr_reg;
+    mcontext_to_user_regs(&mcontext, regs);
+
+    ELF_NOTE_HEADER_TYPE nhdr;
+    // Add one to include the terminating null character.
+    nhdr.n_namesz = strlen(NOTE_OWNER) + 1;
+    nhdr.n_descsz = sizeof(prstatus);
+    nhdr.n_type = NT_PRSTATUS;
+    if (os_write(elf_file, &nhdr, sizeof(nhdr)) != sizeof(nhdr)) {
+        return false;
+    }
+    if (os_write(elf_file, NOTE_OWNER, NOTE_OWNER_LENGTH) != NOTE_OWNER_LENGTH) {
+        return false;
+    }
+    return os_write(elf_file, &prstatus, sizeof(prstatus)) == sizeof(prstatus);
+}
+
+/*
+ * Copy register values from dr_mcontext_t to elf_fpregset_t for AARCH64. For
+ * X86, copy the floating point registers from the output of fxsave64.
+ */
+static bool
+get_floating_point_registers(DR_PARAM_IN dr_mcontext_t *mc,
+                             DR_PARAM_OUT elf_fpregset_t *regs)
+{
+#ifdef DR_HOST_NOT_TARGET
+    return false;
+#elif defined(X86)
+    byte fpstate_buf[MAX_FP_STATE_SIZE];
+    fxsave64_map_t *fpstate = (fxsave64_map_t *)ALIGN_FORWARD(fpstate_buf, 16);
+    if (proc_save_fpstate((byte *)fpstate) != DR_FPSTATE_BUF_SIZE) {
```

Comment:
This won't work for writing out the state of other threads.  See above where we may have to pass in the mcontext.  Add a comment and maybe an assert that the target thread is the executing thread (so maybe want to take in dcontext instead of mcontext).

Reply:
I've changed the code to pass in mcontext from dr_create_memory_dump(), and a TODO to save register values of all threads. PTAL.